### PR TITLE
Fixes bug #34: Stuck in adding seeds

### LINKED ISSUE #34: Stuck in adding seeds
Hi,
Thanks for our sharing. I'm trying to use Bundler+PMVS workflow,  the console seem to be stuck in like below. How could I solve this problem? 

pan@T830:~/bundler_sfm/examples/ET$ ../../bin/genOption pmvs/
pan@T830:~/bundler_sfm/examples/ET$ ../../bin/pmvs2 pmvs/ option-0000

../../bin/pmvs2
pmvs/
option-0000
--------------------------------------------------
--- Summary of specified options ---
# of timages: 4 (enumeration)
# of oimages: 0 (enumeration)
level: 1  csize: 2
threshold: 0.7  wsize: 7
minImageNum: 3  CPU: 8
useVisData: 1  sequence: -1
--------------------------------------------------
Reading images: ****
1 Harris running ...2 Harris running ...3 Harris running ...4 Harris running ...227 harris done
201 harris done
DoG running...DoG running...211 harris done
208 harris done
DoG running...DoG running...315 dog done
314 dog done
315 dog done
316 dog done
done
adding seeds

### PR DESCRIPTION
**Pull request by mostlyuseful**
Fixes #34

Raise language level to C++17 to use std::optional
Replace RWMutex vectors with std::mutex vectors
Replace voluntary locks with optional lock_guards for RAII goodness.

### CODE REVIEW

**Comment by mostlyuseful** [comment]
Clang 5 doesn't have `<optional>` header yet. I'll work around it.

**Comment by mostlyuseful** [comment]
No idea what's going on with Travis CI: I removed `#include <optional>` and pushed, but it still compiles the old version.

File: program/base/pmvs/filter.cc [hunk 1]
```diff
@@ -98,18 +98,19 @@ float Cfilter::computeGain(const Patch::Cpatch& patch, const int lock) {
     const int index2 = iy * m_fm.m_pos.m_gwidths[index] + ix;
     
     float maxpressure = 0.0f;
+    {
+    std::unique_ptr<std::lock_guard<std::mutex>> guard;
```
Comment (pmoulon) [inline program/base/pmvs/filter.cc]: Can you try the reference_wrapper I think it would be more elegant than using unique_ptr?

**Comment by pmoulon** [comment]
Thank you @mostlyuseful PR tested with success.
Would you be willing to continue the update and use c++11 mutex at the remaining stage where the library tiny thread is used?

**Pull request merged**